compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -27,26 +27,147 @@ pub enum AccessLevel {
+            AccessLevel::ReachableFromImplTrait => &mut self.reachable_from_impl_trait,
+        }
+    }
+
+    fn update(&mut self, effective_vis: Visibility, tag: AccessLevel, tree: impl DefIdTree) {
+        for level in [
+            AccessLevel::Public,
+            AccessLevel::Exported,
+            AccessLevel::Reachable,
+            AccessLevel::ReachableFromImplTrait,
+        ] {
+            if level <= tag {
+                let current_effective_vis = self.get_mut(level);
+                if let Some(current_effective_vis) = current_effective_vis {
+                    if current_effective_vis.is_at_least(effective_vis, tree) {
+                        *current_effective_vis = effective_vis;
+                    }
+                } else {
+                    *self.get_mut(level) = Some(effective_vis);
+                }
+            }
+        }
+    }
+
+    fn is_public(&self, tag: AccessLevel) -> bool {
+        self.get(tag).map_or(false, |vis| vis.is_public())
+    }
+}
+
 /// Holds a map of accessibility levels for reachable HIR nodes.
 #[derive(Debug, Clone)]
 pub struct AccessLevels<Id = LocalDefId> {
-    pub map: FxHashMap<Id, AccessLevel>,
+    map: FxHashMap<Id, EffectiveVisibility>,
 }
 
-impl<Id: Hash + Eq> AccessLevels<Id> {
+impl<Id> AccessLevels<Id>
+where
+    Id: Hash + Eq + Copy,
+{
+    fn is_public_at_level(&self, id: Id, tag: AccessLevel) -> bool {
+        self.get_effective_vis(id).map_or(false, |effective_vis| effective_vis.is_public(tag))
+    }
+
     /// See `AccessLevel::Reachable`.
     pub fn is_reachable(&self, id: Id) -> bool {
-        self.map.get(&id) >= Some(&AccessLevel::Reachable)
+        self.is_public_at_level(id, AccessLevel::Reachable)
     }
 
     /// See `AccessLevel::Exported`.
     pub fn is_exported(&self, id: Id) -> bool {
-        self.map.get(&id) >= Some(&AccessLevel::Exported)
+        self.is_public_at_level(id, AccessLevel::Exported)
     }
 
     /// See `AccessLevel::Public`.
     pub fn is_public(&self, id: Id) -> bool {
-        self.map.get(&id) >= Some(&AccessLevel::Public)
+        self.is_public_at_level(id, AccessLevel::Public)
+    }
+
+    pub fn is_some(&self, id: Id) -> bool {
+        self.is_public_at_level(id, AccessLevel::ReachableFromImplTrait)
+    }
+
+    pub fn get_access_level(&self, id: Id) -> Option<AccessLevel> {
+        self.get_effective_vis(id).and_then(|effective_vis| {
+            for level in [
+                AccessLevel::Public,
+                AccessLevel::Exported,
+                AccessLevel::Reachable,
+                AccessLevel::ReachableFromImplTrait,
+            ] {
+                if effective_vis.is_public(level) {
+                    return Some(level);
+                }
+            }
+            None
+        })
+    }
+
+    pub fn set_access_level(
+        &mut self,
+        id: Id,
+        access_level: Option<AccessLevel>,
+        tree: impl DefIdTree,
+    ) -> Option<AccessLevel> {
```

Comment:
```suggestion
    ) {
```
Looks like the returned value is never used.